include/tilck/kernel/paging.h

```diff
@@ -32,9 +32,18 @@
  * These MACROs convert addresses to/from the linear mapping at BASE_VA to the
  * physical address space.
  */
-#define PA_TO_LIN_VA(pa) ((void *) ((ulong)(pa) + BASE_VA))
-#define LIN_VA_TO_PA(va) ((ulong)(va) - BASE_VA)
+#if defined(__i386__) || defined(__x86_64__)
 
+   #define PA_TO_LIN_VA(pa) ((void *) ((ulong)(pa) + BASE_VA))
+   #define LIN_VA_TO_PA(va) ((ulong)(va) - BASE_VA)
+
+#elif defined(__riscv)
+
+   extern ulong linear_va_pa_offset;
```

Comment:
On the i386, the memory physical address must start at 0, but on riscv and other embedded chips, the memory physical address usually does not start at 0, for example, riscv memory physical address usually starts at 0x80000000

Reply:
`va_pa_offset` is calculated like this:

In function `void *init_early_mapping(ulong fdt_paddr)`
```
base_pa = kernel_pa - text_offset; //base_pa indicates the physical start address of the memory
kernel_va_pa_offset = KERNEL_BASE_VA - base_pa;
linear_va_pa_offset = BASE_VA - base_pa;
```

Reply:
Oh yeah, I fully agree on having support for a non-fixed physical start address for the kernel. That's indeed required even on x86-64 for UEFI systems. Tilck doesn't support that yet, however. The kernel needs to be compiled with -fPIC etc.

However, while I completely agree on having this dynamic `linear_va_pa_offset` I'm still confused on the calculations:

```
   ulong kernel_pa = (ulong)&_start - 0x1000;
   ulong text_offset = ((struct linux_image_h *)(void *)&_start)->text_offset;
   ulong base_pa = kernel_pa - text_offset;

   kernel_va_pa_offset = KERNEL_BASE_VA - base_pa;
   linear_va_pa_offset = BASE_VA - base_pa;
 ```
 
 
 So `kernel_pa` is calculated by reading the address of `_start`, from which I deduce that such address is not fixed, right? That means the the kernel should be relocatable, but in the CMake file I see:
 
 ```
 list(
   APPEND ACTUAL_KERNEL_ONLY_FLAGS_LIST

   -D__ACTUAL_TILCK_KERNEL_BINARY__
   -fno-pic

   -fno-common
   -fno-exceptions
   -fno-stack-protector
   -fno-asynchronous-unwind-tables
 ```
 
 So you compile it with `-fno-pic`. I don't get that.
 
Also, wouldn't `&_start` be resolved as a virtual address, even if you didn't enable paging on the CPU yet? If the address of all the symbols where visible from the C code as physical addresses, nothing would work after you enable paging. I must be missing multiple things in your code.

```
ulong text_offset = ((struct linux_image_h *)(void *)&_start)->text_offset;
```

So you're reading the offset of .text from the beginning of _start, which is a struct:

```
#if __riscv_xlen == 64
   /* Image load offset(2MB) from start of RAM */
   .dword 0x200000
#else
   /* Image load offset(4MB) from start of RAM */
   .dword 0x400000
#endif
```

But there we have hard-coded values, not something patched at runtime by the bootloader, right? If those values are really constants, then they should be hard-coded in the main CMakeListst.txt file where we have `BASE_VA` etc. Indeed, that should be a function of `KERNEL_PADDR`. I think that if you define:

```
set(PA_START                  0x80000000) # Can we hard-code that and be generic?
```

Then you could define `KERNEL_PA_OFFSET` as `0x200000` and finally `KERNEL_PADDR` as `PA_START + KERNEL_PA_OFFSET`.

In other words, I don't see how `linear_va_pa_offset` can truly change at runtime with the current implementation as it seems like those calculations are all hard-coded anyway. Is that correct? 

However, if you plan to remove the hard-coded values from start.S / CMake and make the kernel relocatable such that `_start` can actually be anywhere in the physical address space, after PA_START, then it could make sense to read at runtime the address of `_start`. Also:

```
ulong base_pa = kernel_pa - text_offset;
```

Isn't here the implicit assumption is that the kernel is placed at the lowest possible physical address possible? Is that a generic assumption to make for RISCV? 

Sorry for the many questions here, but the memory layout is a fundamental aspect for me to understand.

Reply:
**Just to clarify**: if this is a step towards have a fully relocatable kernel on RISCV64, that would be amazing! I'd also follow the same pattern (whatever is portable of that) on x86.

I'd just like to understand what's truly fixed and what's fixed _for now_ but will be changed by the bootloader later. Linux uses a variable for those calculations because the kernel is indeed fully relocatable on most architectures.

Reply:
> So you compile it with `-fno-pic`. I don't get that.
> 
> Also, wouldn't `&_start` be resolved as a virtual address? If the address of all the symbols were visible from the C code as physical addresses, nothing would work after you enable paging with anything else other than identity mapping. I guess I must be missing multiple things in your code.

Here indeed is a confusing point. When compiling riscv programs, the gcc option `-mcmodel=medany` is usually used(so does tilck). This option means
```
Generate code for the medium-any code model. The program and its statically defined symbols must be within any 
single 2 GiB address range. Programs can be statically or dynamically linked.
The code generated by the medium-any code model is **position-independent**,  but is not guaranteed to function 
correctly when linked into position-independent executables or libraries.
```
Although tilck is compiled with -fno-pic, the code can be considered position-independent until the virtual address is entered. And the value of `&_start` is relative to pc! Therefore before entering the virtual address, we can get the physical address through `&_start` at run time.

> 
> ```
> ulong text_offset = ((struct linux_image_h *)(void *)&_start)->text_offset;
> ```
> 
> So you're reading the offset of `.text` from the beginning of `_start`, which is a struct:
> 
> ```
> #if __riscv_xlen == 64
>    /* Image load offset(2MB) from start of RAM */
>    .dword 0x200000
> #else
>    /* Image load offset(4MB) from start of RAM */
>    .dword 0x400000
> #endif
> ```
> 
> But there we have hard-coded values, not something patched at runtime by the bootloader, right? If those values are really constants, then they should be hard-coded in the main CMakeListst.txt file where we have `BASE_VA` etc. Indeed, those values should be a function of `KERNEL_PADDR`. If you define:
> 
> ```
> set(PA_START                  0x80000000) # Can we hard-code that and be generic?
> ```
>
> Then you could define `KERNEL_PA_OFFSET` as `0x200000` and finally `KERNEL_PADDR` as `PA_START + KERNEL_PA_OFFSET`.

We can not define those values in CMake file, because U-boot (Or any other bootloaders used to boot linux) will read these variables written in the kernel's head.
In addition, the value of `text_offset` is indeed a constant, However, we cannot assume that PA_START is always equal to 0x80000000, and this value is really different on different models of chips.

> In other words, I don't see how `linear_va_pa_offset` can truly change at runtime with the _current implementation_ as it seems like those calculations are all hard-coded anyway. Is that correct?
> 

See what I explained to `&_start` above.

> However, if you plan to remove the hard-coded values from start.S / CMake and make the kernel relocatable such that `_start` can actually be anywhere in the physical address space, after PA_START, then it could make sense to read at runtime the address of `_start`. Also:
> 
> ```
> ulong base_pa = kernel_pa - text_offset;
> ```
> 
> Isn't here the implicit assumption is that the kernel is placed at the lowest possible physical address possible? Is that a generic assumption to make for RISCV?
> 

Yes, that is a generic assumption on RISCV. In addition, when kernel is running, at the really beginning of the physical address is the sbi program running in machine mode(This address space is also protected and cannot be touched by kernel), and the kernel calls sbi through s-mode ecall(similar to user space syscall)

Reply:
> **Just to clarify**: if this is a step towards have a fully relocatable kernel on RISCV64, that would be amazing! I'd also follow the same pattern (whatever is portable of that) on x86.
> 
> I'd just like to understand what's truly fixed and what's fixed _for now_ but will be changed by the bootloader later. Linux uses a variable for those calculations because the kernel is indeed fully relocatable on most architectures.

Yes，I tried to make tilck fully relocatable on RISCV64. But I also want to point out that because tilck is loaded into memory in **elf format**, loading program segments still relies on pre-defined physical addresses. Of course, this can be fixed by modifying the function which loads elf in U-boot, or by other method in the future.